### PR TITLE
utils: port hooks to Python

### PR DESCRIPTION
**Pull request by mbouron-gpsw**

### CODE REVIEW

**Comment by jmoguillansky-gpsw** [comment]
Please test on Windows

**Comment by mbouron-gpsw** [comment]
> Please test with ngl-desktop/ngl-control on Windows

ngl-control does not work under Windows currently. This two commits have been tested on Windows (MSVC) by using the hooks manually.

**Review by jmoguillansky-gpsw — COMMENTED** [review]
> > Please test with ngl-desktop/ngl-control on Windows
> 
> ngl-control does not work under Windows currently. This two commits have been tested on Windows (MSVC) by using the hooks manually.

It works in multibackend branch.  Can you test this code there?

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_session_info [hunk 1]
```diff
@@ -1,10 +1,9 @@
-#!/bin/sh
+#!/usr/bin/env python
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_session_info:1]: Calling directly external python code means that the mobile hooks needs to be ported to python and I'm not sure yet if that's feasible for iOS.

[the remark this answers is not in the document]
Comment (jmoguillansky-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_session_info:1]: I see

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
-for session in $NGL_DESKTOP_REMOTE_SESSIONS; do
-    host=${session%%-*}
-    port=${session##*-}
-    if ngl-ipc -x $host -p $port >/dev/null; then
-        echo $session remote ngl-desktop
-    fi
-done
+tmp_dir = os.path.join(tempfile.gettempdir(), 'ngl-desktop')
+if not os.path.isdir(tmp_dir):
+    sys.exit(0)
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions]: I rewrote this part so we don't exit too early and it makes the code able to reach the remote session part. It doesn't require more error checking on this particular part though. It is ok for the tmp directory to not exist, it just means that ngl-desktop has not been started on this machine.

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_session_info [hunk 3]
```diff
@@ -1,10 +1,9 @@
...
 
-session=$1
-host=${session%%-*}
-port=${session##*-}
-
-set -x
-ngl-ipc -x "$host" -p "$port" -?
+session = sys.argv[1]
+host, port = session.rsplit('-', maxsplit=1)
+ret = subprocess.call(['ngl-ipc', '-x', host, '-p', port, '-?'])
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_session_info:8]: We won't see the commands anymore: `set -x` causes commands to be displayed the command on stderr, which can be useful to monitor the communication going on, as well as debugging (typically to re-exec the command manually). Not a exactly a blocking comment, but this is not an improvement IMO.

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions [hunk 4]
```diff
@@ -1,21 +1,32 @@
...
+
+session_dirs = []
+if os.path.isdir(tmp_dir):
+    session_dirs = os.listdir(tmp_dir)
+for session_dir in session_dirs:
+    session_file = os.path.join(tmp_dir, session_dir, 'session')
+    if not os.path.isfile(session_file):
+        continue
+    host, port = session_dir.rsplit('-', maxsplit=1)
+    ret = subprocess.call(['ngl-ipc', '-x', host, '-p', port], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions:19]: only stdout needs to be filtered out: stderr is actually useful for diagnosis

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions [hunk 5]
```diff
@@ -1,21 +1,32 @@
...
+    if ret == 0:
+        print(f'{session_dir} local ngl-desktop')
+
+remote_sessions = os.getenv('NGL_DESKTOP_REMOTE_SESSIONS', None)
+if not remote_sessions:
+    sys.exit(0)
+for session in remote_sessions.split(' '):
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions:26]:
```python
remote_sessions = os.getenv('NGL_DESKTOP_REMOTE_SESSIONS', '').split()
for session in remote_sessions:
    ...
```

Make sure not to explicit the `' '` to split, it's less flexible than the default:
```
>>> 'a   b'.split(' ')
['a', '', '', 'b']
>>> 'a   b'.split()
['a', 'b']
```

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions [hunk 6]
```diff
@@ -1,21 +1,32 @@
...
+    host, port = session.rsplit('-', maxsplit=1)
+    ret = subprocess.call(['ngl-ipc', '-x', host, '-p', port], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
+    if ret == 0:
+        print(f'{session_dir} remote ngl-desktop')
+
+sys.exit(0)
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_sessions:32]: this looks unnecessary

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.scene_change [hunk 7]
```diff
@@ -1,22 +1,24 @@
...
+    'aspect_ratio': '-a',
+    'framerate':    '-r',
+    'clear_color':  '-c',
+    'samples':      '-m',
+}
 
-set -x
+cmd = ['ngl-ipc', '-x', host, '-p', port]
+for arg in sys.argv[3:]:
+    name, value = arg.rsplit('=', maxsplit=1)
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hook.scene_change:20]: no need for `rsplit` here, `split` is enough

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.sync_file [hunk 8]
```diff
@@ -1,19 +1,16 @@
...
+import sys
+import subprocess
 
-session=$1
-ifile=$2
-ofile=$3
+session = sys.argv[1]
+host, port = session.rsplit('-', maxsplit=1)
+ifile = sys.argv[2]
+ofile = sys.argv[3]
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hook.sync_file:9]: you can do stuff like `session, ifile, ofile = sys.argv[1:4]`

**Review by cboesch-gpsw — COMMENTED** [review]
Overall I don't really mind (except the loss of the log tracing that bugs me a bit), but since it doesn't directly help the Windows support (which seems to require transforming the executables/scripts into a Python API and not just Python scripts), I'm not so sure about it?

**Comment by jmoguillansky-gpsw** [comment]
> Overall I don't really mind (except the loss of the log tracing that bugs me a bit), but since it doesn't directly help the Windows support (which seems to require transforming the executables/scripts into a Python API and not just Python scripts), I'm not so sure about it?

Windows support doesn't require transforming the scripts into a Python API?  
My initial attempt to port ngl-control/ngl-desktop involved reusing the bash scripts but changing the path params.
I'm ok with these changes and switching to python scripts, as long as the functionality works.  Its up to you guys

**Comment by jmoguillansky-gpsw** [comment]
my only suggestion is that it's nice to be able to run the hooks manually, with or without the gui, but that's already supported with this change, and I think we're on the same page?

**Comment by mbouron-gpsw** [comment]
> my only suggestion is that it's nice to be able to run the hooks manually, with or without the gui, but that's already supported with this change, and I think we're on the same page?

Yes, the hooks can still be executed without ngl-control or any GUI.

File: pynodegl-utils/pynodegl_utils/hooks.py [hunk 9]
```diff
@@ -89,15 +78,14 @@ def _uint_clear_color(vec4_color):
         return uint_color
 
     def scene_change(self, session_id, local_scene, cfg):
-        self._get_hook_output(
-            'scene_change',
+        return self._module.scene_change(
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks.py:81]: `_get_hook_output()` (and other functions) were resilient to missing hooks (relying on `hooks_available`). It looks like this is not the case anymore.

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import os
+import sys
+import subprocess
+import tempfile
+
+
+def get_session_info(session_id):
+    host, port = session_id.rsplit('-', maxsplit=1)
+    output = subprocess.check_output(['ngl-ipc', '-x', host, '-p', port, '-?'], text=True).rstrip()
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py:9]: what is `text=True` for?

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    ret = {}
+    for line in output.splitlines():
+        k, v = line.split('=', maxsplit=1)
+        ret[k] = v
+    return ret
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py:14]: Mmh... `return dict(line.split('=', maxsplit=1) for line in output.splitlines())`?

File: pynodegl-utils/pynodegl_utils/hooks.py [hunk 12]
```diff
@@ -34,44 +35,32 @@ class _HooksCaller:
...
-    def _get_hook_output(self, name, *args):
-        hook = self._get_hook(name)
-        if not hook:
-            return None
-        cmd = [hook] + list(args)
-        return subprocess.check_output(cmd, text=True).rstrip()
+    def _has_hook(self, name):
+        if hasattr(self._module, name):
+            return True
+        return False
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks.py:52]: Done.

File: doc/ref/pynodegl-utils.md [hunk 13]
```diff
@@ -192,90 +192,82 @@ def demo(cfg, intro='Hello World!'):
...
+7. `samples`: a `int` representing the number of `samples` used for multisample anti-aliasing
 
 **Example**:
 
-A call from `ngl-control` to this hook will look like this:
-
-```shell
-$ ./hook.scene_change X2fca1f2c /tmp/ngl_scene.ngl duration=5 framerate=60000/1001 aspect_ratio=16/9 clear_color=4A646BFF samples=4
+```console
+$ python -c 'import pprint, hooks; pprint.pprint(hooks.sync_file("192.168.1.1-1234", "/tmp/scene.ngl", "5", "60000/1001", "16/9", "4A646BFF", "4"))'
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply doc/ref/pynodegl-utils.md]: Fixed.

File: doc/ref/pynodegl-utils.md [hunk 14]
```diff
@@ -192,90 +192,82 @@ def demo(cfg, intro='Hello World!'):
...
 
-The hook must print on `stdout` the output file path.
+The hook must return the output file path.
 
 **Example**:
 
-```shell
-$ ./hook.sync_file Y5fd953df /tmp/ngl-media.mp4 media-001.mp4
+```console
+$ python -c 'import pprint, hooks; pprint.pprint(hooks.sync_file("192.168.1.1-1234", "/tmp/ngl-media.mp4", "media-001.mp4"))'
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply doc/ref/pynodegl-utils.md]: Fixed.

File: doc/ref/pynodegl-utils.md [hunk 15]
```diff
@@ -192,90 +192,82 @@ def demo(cfg, intro='Hello World!'):
...
-### hook.sync_file
+### sync_file
 
-`hook.sync_file` takes 3 arguments:
+`sync_file(session_id, ifile, ofile)` takes the following arguments:
 
 1. `session_id`: the session identifier
-1. `ifile`: the path to the local file
+2. `ifile`: the path to the local file
 2. `ofile`: the output filename (not path)
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply doc/ref/pynodegl-utils.md:245]: I prefer to have a file that is more human readable as it's easier to work with.

File: doc/ref/pynodegl-utils.md [hunk 16]
```diff
@@ -192,90 +192,82 @@ def demo(cfg, intro='Hello World!'):
...
 
 **Example**:
 
-```shell
-$ ./hook.get_session_info X2fca1f2c
-backend=opengles
-system=Linux
+```console
+$ python -c 'import pprint, hooks; pprint.pprint(hooks.get_session_info("localhost-1234"))'
+{'backend': 'opengl', 'system': 'Linux'}
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply doc/ref/pynodegl-utils.md]: Fixed.

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py (hunk 11 above)
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py:14]: Done.

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py (hunk 10 above)
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py:9]: without text=True we got the following exception:
```
Traceback (most recent call last):
  File "/home/mateo/src/devel/node.gl/pynodegl-utils/pynodegl_utils/ui/hooks_view.py", line 150, in _refresh
    for row_id, (session_id, desc, backend, system) in enumerate(self._hooks_caller.get_sessions()):
  File "/home/mateo/src/devel/node.gl/pynodegl-utils/pynodegl_utils/hooks.py", line 127, in get_sessions
    caller_sessions = caller.get_sessions()
  File "/home/mateo/src/devel/node.gl/pynodegl-utils/pynodegl_utils/hooks.py", line 61, in get_sessions
    session_info = self._module.get_session_info(session_id)
  File "/home/mateo/src/devel/node.gl/pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py", line 10, in get_session_info
    return dict(line.split('=', maxsplit=1) for line in output.splitlines())
  File "/home/mateo/src/devel/node.gl/pynodegl-utils/pynodegl_utils/hooks/desktop/hooks.py", line 10, in <genexpr>
    return dict(line.split('=', maxsplit=1) for line in output.splitlines())
TypeError: a bytes-like object is required, not 'str'
```

File: pynodegl-utils/pynodegl_utils/hooks.py [hunk 17]
```diff
@@ -34,44 +35,32 @@ class _HooksCaller:
...
-        self._hooksdir = hooksdir
-        self.hooks_available = all(self._get_hook(h) is not None for h in self._HOOKS)
-        if not self.hooks_available:
-            print(f'hooks not available in {hooksdir}')
+    def __init__(self, hooks_script):
+        name, _ = op.splitext(op.basename(hooks_script))
+        spec = importlib.util.spec_from_file_location(name, hooks_script)
+        module = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(module)
+        self._module = module
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks.py]: Done.

File: pynodegl-utils/pynodegl_utils/hooks.py (hunk 9 above)
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks.py:74]: The _HookCaller now error out early if all the hooks are not implemented.

File: pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_session_info (hunk 3 above)
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/hooks/desktop/hook.get_session_info]: Done by adding a wrapper on top of subprocess in the hook script.

File: doc/ref/pynodegl-utils.md [hunk 18]
```diff
@@ -192,90 +192,82 @@ def demo(cfg, intro='Hello World!'):
...
-  - `framerate`: expressed as a fraction in format `num/den` where `num` and `den` are `int`
-  - `clear_color`: expressed as a 32-bit hexadecimal following the `RRGGBBAA` format
-  - `samples`: number of samples used for multisample anti-aliasing expressed as an integer
+`scene_change(session_id, scenefile, duration, aspect_ratio, framerate, clear_color, samples)` take the following arguments:
+1. `session_id`: the session identifier
+2. `scenefile`: the path to the local serialized scene
+3. `duration`: a `float` representing the scene duration (in seconds)
+4. `aspect_ratio`: a `tuple` of `int` (`num`, `den`) representing the aspect ratio
+5. `framerate`: a `tuple` of `int` (`num`, `den`) representing the framerate
+6. `clear_color`: a string representing a color expressed as a 32-bit hexadecimal following the `RRGGBBAA` format
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply doc/ref/pynodegl-utils.md]: Done.

File: doc/ref/pynodegl-utils.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply doc/ref/pynodegl-utils.md]: Fixed.

File: pynodegl-utils/pynodegl_utils/controller.py [hunk 19]
```diff
@@ -34,13 +34,13 @@ def run():
     parser = argparse.ArgumentParser()
     parser.add_argument('-m', dest='module', default='pynodegl_utils.examples',
                         help='set the module name containing the scene functions')
-    parser.add_argument('--hooks-dir', dest='hooksdirs',
-                        default=[op.join(op.dirname(__file__), 'hooks', 'desktop')],
+    parser.add_argument('--hooks-scripts', dest='hooks_scripts',
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/controller.py]: Fixed.

**Review by cboesch-gpsw — APPROVED** [review]

**Pull request merged**